### ISSUE TITLE
SQLFluff incorrectly lowercased JSON attributes causing significant outage [Snowflake]

### ISSUE THREAD

**Issue by jlucas91** [sqlfluff/sqlfluff#5659]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

We store json blobs within columns throughout our data pipeline. We'll then parse the JSON blob's field/values into their own columns in later stages. SQLFluff incorrectly lowercased the access path when retrieving data from the JSON blob. This lead to all accesses silently failing. 

For example, given the original code:
```
select
    ...
    cast(address:streetAddress1 as varchar) as address_line1,
    ...
from source
```

SQL fluff converted `streetAddress1` to `streetaddress1`. 

### Expected Behaviour

Lowercasing access paths within variant object access is not safe and should not occur. 

### Observed Behaviour

All access paths were lowercased. 

### How to reproduce

Assumedly, this is a consistent behavior for most people on snowflake?

### Dialect

Snowflake

### Version

3.0.0a5

### Configuration

[sqlfluff]
dialect = snowflake
templater = dbt
max_line_length = 88
runaway_limit = 10
indent_unit = space
exclude_rules = layout.indent, layout.cte_bracket, layout.cte_newline, layout.select_targets, jinja.padding, references.keywords, aliasing.self_alias.column
processes = -1

[sqlfluff:indentation]
tab_space_size = 4

[sqlfluff:layout:type:comma]
spacing_before = touch
line_position = trailing

[sqlfluff:rules:capitalisation.keywords] 
capitalisation_policy = lower

[sqlfluff:rules:aliasing.table]
aliasing = explicit

[sqlfluff:rules:aliasing.column]
aliasing = explicit

[sqlfluff:rules:aliasing.expression]
allow_scalar = False

[sqlfluff:rules:capitalisation.identifiers]
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower

[sqlfluff:rules:ambiguous.column_references]  # Number in group by
group_by_and_order_by_style = implicit

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Hi @jlucas91 this definitely shouldn't happen, and I'm having trouble recreating your issue.

If I parse your sample query:

```sql
select
    cast(address:streetAddress1 as varchar) as address_line1
from source
```

Then this is the parse output:

```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                newline:                                      '\n'
[L:  2, P:  1]      |                whitespace:                                   '    '
[L:  2, P:  5]      |                select_clause_element:
[L:  2, P:  5]      |                    function:
[L:  2, P:  5]      |                        function_name:
[L:  2, P:  5]      |                            function_name_identifier:         'cast'
[L:  2, P:  9]      |                        bracketed:
[L:  2, P:  9]      |                            start_bracket:                    '('
[L:  2, P: 10]      |                            [META] indent:
[L:  2, P: 10]      |                            expression:
[L:  2, P: 10]      |                                column_reference:
[L:  2, P: 10]      |                                    naked_identifier:         'address'
[L:  2, P: 17]      |                                semi_structured_expression:
[L:  2, P: 17]      |                                    colon:                    ':'
[L:  2, P: 18]      |                                    semi_structured_element:  'streetAddress1'
[L:  2, P: 32]      |                            whitespace:                       ' '
[L:  2, P: 33]      |                            keyword:                          'as'
[L:  2, P: 35]      |                            whitespace:                       ' '
[L:  2, P: 36]      |                            data_type:
[L:  2, P: 36]      |                                data_type_identifier:         'varchar'
[L:  2, P: 43]      |                            [META] dedent:
[L:  2, P: 43]      |                            end_bracket:                      ')'
[L:  2, P: 44]      |                    whitespace:                               ' '
[L:  2, P: 45]      |                    alias_expression:
[L:  2, P: 45]      |                        [META] indent:
[L:  2, P: 45]      |                        keyword:                              'as'
[L:  2, P: 47]      |                        whitespace:                           ' '
[L:  2, P: 48]      |                        naked_identifier:                     'address_line1'
[L:  2, P: 61]      |                        [META] dedent:
[L:  2, P: 61]      |                [META] dedent:
[L:  2, P: 61]      |            newline:                                          '\n'
[L:  3, P:  1]      |            from_clause:
[L:  3, P:  1]      |                keyword:                                      'from'
[L:  3, P:  5]      |                whitespace:                                   ' '
[L:  3, P:  6]      |                from_expression:
[L:  3, P:  6]      |                    [META] indent:
[L:  3, P:  6]      |                    from_expression_element:
[L:  3, P:  6]      |                        table_expression:
[L:  3, P:  6]      |                            table_reference:
[L:  3, P:  6]      |                                naked_identifier:             'source'
[L:  3, P: 12]      |                    [META] dedent:
[L:  3, P: 12]      |    newline:                                                  '\n'
[L:  4, P:  1]      |    [META] end_of_file:
```

You can see that the `streetAddress1` element is parsed as a `semi_structured_element`, which shouldn't be affected by any of the capitalisation rules. Can you narrow down which rule is causing your issue and a full query which recreates the issue?

**Comment by jlucas91** [comment]
Ack - so sorry for the distraction @alanmcruickshank . On further review, it appears this was sqlfmt, not sqlfluff that caused the problem. We had been evaluating the pros/cons of the two linters and sqlfmt had pre-processed this file before sqlfluff got to it. This has been one of those postmortems where 10 things all go wrong at the same time 🤦 . Appreciate your time and review.

**Comment by alanmcruickshank** [comment]
No stress. I'm glad that SQLFluff won the comparison! 😁 

Thanks for following up 👍
